### PR TITLE
First version of the `tls-requirer` operator

### PR DESCRIPTION
**Pull request by gruyaume**
## Overview
First version of the `tls-requirer` operator. Charm that requests X.509 certificates using the `tls-certificates` interface.

It uses the user-provided `subject` configuration to generate a certificate signing request
(CSR) that will be inserted into its unit relation data as soon as the `tls-certificates` relation
is created. In return, the certificate provider should use this CSR, generate a certificate, 
and provide it back into their application relation data.

This charm is useful when developing and testing certificate providers.

## Usage

Deploy the charm and relate it to a certificate provider:

```bash
juju deploy tls-requirer-operator
juju relate tls-requirer-operator <TLS Certificates Provider>
```

Access the generated certificate:

```bash
juju run tls-requirer-operator/leader get-certificate
```

## Limitations

This charm doesn't scale up.

## Relations

- `tls-certificates`: Used for charms that require/provide TLS certificates.

### CODE REVIEW

File: .github/workflows/publish-charm.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      - name: Chmod charmcraft logs
+        if: steps.upload.conclusion == 'failure'
+        run: |
+          sudo chmod -R 777 /root/snap/charmcraft/common/cache/charmcraft/log/
+      - name: Archive charmcraft logs
+        if: steps.upload.conclusion == 'failure'
+        uses: actions/upload-artifact@v3
+        with:
+          name: charmcraft-upload-logs
+          path: /root/snap/charmcraft/common/cache/charmcraft/log/*.log
```
Comment (ghislainbourgeois) [inline .github/workflows/publish-charm.yaml:36]: I think this is not required

File: config.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+options:
+  subject:
+    type: string
+    default: ubuntu.com
```
Comment (ghislainbourgeois) [inline config.yaml:7]: I don't think we want to use `ubuntu.com` here.
```suggestion
    default: tls-requirer-operator
```

File: src/charm.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        Args:
+            relation_name (str): Relation name
+
+        Returns:
+            bool: Whether a given relation was created.
+        """
+        try:
+            if self.model.get_relation(relation_name):
+                return True
+            return False
```
Comment (ghislainbourgeois) [inline src/charm.py:89]:
```suggestion
            return bool(self.model.get_relation(relation_name))
```

File: src/charm.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+
+def generate_password() -> str:
+    """Generates a random 12 character password.
+
+    Returns:
+        str: Password
+    """
+    chars = string.ascii_letters + string.digits
+    return "".join(secrets.choice(chars) for _ in range(12))
```
Comment (ghislainbourgeois) [inline src/charm.py:306]:
```suggestion
    return secrets.token_bytes(12)
```

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: .github/workflows/publish-charm.yaml (hunk 1 above)
Comment (gruyaume) [reply .github/workflows/publish-charm.yaml]: Removed

**Review by ghislainbourgeois — APPROVED** [review]

**Pull request merged**